src/hotspot/share/opto/superword.cpp

```diff
@@ -2346,6 +2348,204 @@ void SuperWord::verify_packs() {
+  int _max_pid = 0;
+
+  GrowableArray<Node*> &_block;
+  GrowableArray<Node_List*> &_packset;
+  DepGraph &_dg;
+public:
+  PacksetGraph(Arena* a,
+                GrowableArray<Node*> &block,
+                GrowableArray<Node_List*> &packset,
+                DepGraph &dg)
+  : _pid(a, 8, 0, /* default */ 0), _block(block), _packset(packset), _dg(dg) {
+  }
+  // Get pid, if there is a packset node that n belongs to. Else return 0.
+  int get_pid_or_zero(const Node* n) const {
+    if ((int)n->_idx >= _pid.length()) {
+      return 0;
+    } else {
+      return _pid.at(n->_idx);
+    }
+  }
+  int get_pid(const Node* n) {
+    int poz = get_pid_or_zero(n);
+    assert(poz != 0, "pid should not be zero");
+    return poz;
+  }
+  void set_pid(const Node* n, int pid) {
+    assert(n != nullptr && pid > 0, "sane inputs");
+    _pid.at_put_grow(n->_idx, pid);
+  }
+  int new_pid() {
+    _incnt.push(0);
+    _out.push(GrowableArray<int>());
+    return ++_max_pid;
+  }
+  int incnt(int pid) { return _incnt.at(pid - 1); }
+  void incnt_set(int pid, int cnt) { return _incnt.at_put(pid - 1, cnt); }
+  GrowableArray<int>& out(int pid) { return _out.at(pid - 1); }
+
+  // Create nodes (from packs and scalar-nodes), and add edges, based on DepPreds.
+  void build() {
+    // Map nodes in packsets
+    for (int i = 0; i < _packset.length(); i++) {
+      Node_List* p = _packset.at(i);
+      int pid = new_pid();
+      for (uint k = 0; k < p->size(); k++) {
+        Node* n = p->at(k);
+        set_pid(n, pid);
+      }
+    }
+
+    int max_pid_packset = _max_pid;
+
+    // Map nodes not in packset
+    for (int i = 0; i < _block.length(); i++) {
+      Node* n = _block.at(i);
+      if (n->is_Phi() || n->is_CFG()) {
+        continue; // ignore control flow
+      }
+      int pid = get_pid_or_zero(n);
+      if (pid == 0) {
+        pid = new_pid();
+        set_pid(n, pid);
+      }
+    }
+
+    // Map edges for packset nodes
+    VectorSet set;
+    for (int i = 0; i < _packset.length(); i++) {
+      Node_List* p = _packset.at(i);
+      set.clear();
+      for (uint k = 0; k < p->size(); k++) {
+        Node* n = p->at(k);
+        int pid = get_pid(n);
+        for (DepPreds preds(n, _dg); !preds.done(); preds.next()) {
+          Node* pred = preds.current();
+          int pred_pid = get_pid_or_zero(pred);
+          if (pred_pid == pid && n->is_reduction()) {
+            continue; // reduction -> self-cycle is not a cyclic dependency
+          }
+          // Only add edges once, and only for mapped nodes (in _block)
+          if (pred_pid > 0 && !set.test_set(pred_pid)) {
+            incnt_set(pid, incnt(pid) + 1); // increment
+            out(pred_pid).push(pid);
+          }
+        }
+      }
+    }
+
+    // Map edges for nodes not in packset
+    for (int i = 0; i < _block.length(); i++) {
+      Node* n = _block.at(i);
+      int pid = get_pid_or_zero(n);
+      if (pid == 0 || pid <= max_pid_packset) {
```

Comment:
Is `pid == 0 ||` repetitive here? We have `max_pid_packset >= 0`, right?

Reply:
Good catch.